daprovider/referenceda/reference_reader.go

```diff
@@ -0,0 +1,154 @@
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create validator binding: %w", err)
+		}
+
+		// Validate using contract
+		callOpts := &bind.CallOpts{Context: ctx}
+		err = cert.ValidateWithContract(validator, callOpts)
+		if err != nil {
+			return nil, nil, fmt.Errorf("certificate validation failed: %w", err)
+		}
+	*/
+
+	log.Debug("ReferenceDA reader extracting hash",
+		"certificateLen", len(certBytes),
+		"sha256Hash", common.Hash(cert.DataHash).Hex(),
+		"certificateHex", fmt.Sprintf("0x%x", certBytes))
+
+	// Retrieve the data from storage using the hash
+	var payload []byte
+	if needPayload || needPreimages {
+		payload, err = r.storage.GetByHash(ctx, common.BytesToHash(cert.DataHash[:]))
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to retrieve data from storage: %w", err)
+		}
+		if payload == nil {
+			return nil, nil, fmt.Errorf("data not found in storage for hash %s", common.Hash(cert.DataHash).Hex())
+		}
+
+		// Verify data matches certificate hash (SHA256)
+		actualHash := sha256.Sum256(payload)
+		if actualHash != cert.DataHash {
+			return nil, nil, fmt.Errorf("data hash mismatch: expected %s, got %s", common.Hash(cert.DataHash).Hex(), common.Hash(actualHash).Hex())
+		}
+	}
+
+	// Record preimages if needed
+	var preimages daprovider.PreimagesMap
+	if needPreimages && payload != nil {
+		preimages = make(daprovider.PreimagesMap)
+		preimageRecorder := daprovider.RecordPreimagesTo(preimages)
+
+		// Record the mapping from certificate hash to actual payload data
+		// This is what the replay binary expects: keccak256(certificate) -> payload
+		certHash := crypto.Keccak256Hash(certBytes)
+		preimageRecorder(certHash, payload, arbutil.DACertificatePreimageType)
+	}
+
+	log.Debug("ReferenceDA batch recovery completed",
+		"batchNum", batchNum,
+		"blockHash", batchBlockHash,
+		"sha256", common.Hash(cert.DataHash).Hex(),
+		"payloadSize", len(payload))
+
+	return payload, preimages, nil
+}
+
+// RecoverPayload fetches the underlying payload from the DA provider
+func (r *Reader) RecoverPayload(
+	batchNum uint64,
+	batchBlockHash common.Hash,
+	sequencerMsg []byte,
+) containers.PromiseInterface[daprovider.PayloadResult] {
+	promise, ctx := containers.NewPromiseWithContext[daprovider.PayloadResult](context.Background())
+	go func() {
+		payload, _, err := r.recoverInternal(ctx, batchNum, batchBlockHash, sequencerMsg, true, false)
+		if err != nil {
+			promise.ProduceError(err)
+		} else {
+			promise.Produce(daprovider.PayloadResult{Payload: payload})
+		}
+	}()
+	return promise
+}
+
+// CollectPreimages collects preimages from the DA provider
+func (r *Reader) CollectPreimages(
+	batchNum uint64,
+	batchBlockHash common.Hash,
+	sequencerMsg []byte,
+) containers.PromiseInterface[daprovider.PreimagesResult] {
+	promise, ctx := containers.NewPromiseWithContext[daprovider.PreimagesResult](context.Background())
+	go func() {
```

Comment:
now I think we can also encapsulate further with something like
```
func LaunchPromisethread[T any](parent context.Context, threadfunc func(ctx context) (T, error)) *PromiseInterface[T]
```
not urgent, and we could keep both :)

Reply:
very much like stopwaiter.LaunchPromiseThread, but without the stopwaiter encapsulation